server/eventlog/eventlog.go

```diff
@@ -23,41 +37,243 @@ func GetEventLogChunk(ctx context.Context, env environment.Env, req *elpb.GetEve
+				// id of the next chunk to be written and PreviousChunkId set to the id
+				// of the last chunk.
+				return &elpb.GetEventLogChunkResponse{
+					Chunk: &elpb.GetEventLogChunkResponse_Chunk{
+						Buffer: make([]byte, 0),
+					},
+					PreviousChunkId: chunkstore.ChunkIndexAsString(lastChunkIndex),
+					NextChunkId:     chunkstore.ChunkIndexAsString(lastChunkIndex + 1),
+				}, nil
+			} else {
+				// If out-of-bounds and the invocation is not in-progress, return an
+				// empty chunk with NextChunkId left blank and PreviousChunkId set to
+				// the id of the last chunk.
+				return &elpb.GetEventLogChunkResponse{
+					Chunk: &elpb.GetEventLogChunkResponse_Chunk{
+						Buffer: make([]byte, 0),
+					},
+					PreviousChunkId: chunkstore.ChunkIndexAsString(lastChunkIndex),
+				}, nil
+			}
+		}
 	}
 
 	rsp := &elpb.GetEventLogChunkResponse{
 		Chunk: &elpb.GetEventLogChunkResponse_Chunk{
-			ChunkId: fmt.Sprintf("%04x", intChunkId),
+			ChunkId: chunkstore.ChunkIndexAsString(chunkIndex),
+			Buffer:  make([]byte, 0),
 		},
+		NextChunkId:     chunkstore.ChunkIndexAsString(chunkIndex + 1),
+		PreviousChunkId: chunkstore.ChunkIndexAsString(chunkIndex - 1),
 	}
-	if rsp.Chunk.Buffer, err = c.ReadChunk(ctx, GetEventLogPathFromInvocationId(req.InvocationId), intChunkId); err != nil {
-		return nil, err
+
+	readBackward := false
+	if req.ReadBackward || req.ChunkId == "" {
+		readBackward = true
+	}
+	boundary := lastChunkIndex + 1
+	step := uint16(1)
+	if readBackward {
+		boundary = math.MaxUint16
+		step = math.MaxUint16 // decrements the value when added
+	}
+	// Fetch one chunk even if the minimum line count is 0
+	for lineCount := 0; (rsp.Chunk.ChunkId != "" || lineCount < int(req.MinLines)) && chunkIndex != boundary; chunkIndex += step {
+		buffer, err := c.ReadChunk(ctx, eventLogPath, chunkIndex)
+		if err != nil {
+			return nil, err
+		}
+		scanner := bufio.NewScanner(bytes.NewReader(buffer))
+		for scanner.Scan() {
+			if scanner.Err() != nil {
+				return nil, err
+			}
+			lineCount++
+		}
+		if readBackward {
+			rsp.Chunk.Buffer = append(buffer, rsp.Chunk.Buffer...)
+		} else {
+			rsp.Chunk.Buffer = append(rsp.Chunk.Buffer, buffer...)
+		}
+		if rsp.Chunk.ChunkId != chunkstore.ChunkIndexAsString(chunkIndex) {
+			// No longer fetching a single chunk, ChunkId cannot be meaningfully set.
+			rsp.Chunk.ChunkId = ""
+		}
 	}
-	if intChunkId > 0 {
-		rsp.PreviousChunkId = fmt.Sprintf("%04x", intChunkId-1)
+	if readBackward {
+		rsp.PreviousChunkId = chunkstore.ChunkIndexAsString(chunkIndex)
+	} else {
+		rsp.NextChunkId = chunkstore.ChunkIndexAsString(chunkIndex)
 	}
-	if intChunkId < math.MaxInt16 {
-		rsp.NextChunkId = fmt.Sprintf("%04x", intChunkId+1)
+
+	invalidChunkId := chunkstore.ChunkIndexAsString(math.MaxUint16)
+	if rsp.NextChunkId == invalidChunkId {
+		rsp.NextChunkId = ""
+	}
+	if rsp.PreviousChunkId == invalidChunkId {
+		rsp.PreviousChunkId = ""
 	}
```

Comment:
I think we should return the `invalidChunkId` back to the client so that the client will then request the invalid chunk ID, and then the client will surface an error like "Log storage limit exceeded". Otherwise the client will see that `nextChunkId == ""` and stop fetching, and the user won't get any feedback that something went wrong

Reply:
done.